File: thisweek_masuipeo/q608/tyama_codeiq608_next.java
import java.util.*;
class Main{
	static <T> void reverse(List<T> a,int start,int size){
		int end=start+size-1;
		for(;start<end;start++){
			T z=a.get(start);a.set(start,a.get(end));a.set(end,z);
			end--;
		}
	}
	static <T extends Comparable<? super T>> boolean next_permutation(List<T> a,int n){
		if(n<0||a.size()<n)return false;
		int i;
		reverse(a,n,a.size()-n);
		for(i=a.size()-2;i>=0;i--)if(a.get(i).compareTo(a.get(i+1))<0)break;
		if(i<0){
			reverse(a,0,a.size());
			return false;
		}
		int k=i;
		for(i=a.size()-1;i>=k+1;i--)if(a.get(k).compareTo(a.get(i))<0)break;
		int l=i;
		T z=a.get(k);a.set(k,a.get(l));a.set(l,z);
		reverse(a,k+1,a.size()-(k+1));
		return true;
	}
	static <T extends Comparable<? super T>> boolean next_permutation(List<T> a){
		return next_permutation(a,a.size());
	}

	static final int N=6;
	public static void main(String[]z){
		int r=0,i;
		List<Integer>e0=new ArrayList<Integer>(),f0=new ArrayList<Integer>();
		for(i=0;i<N;i++){e0.add(0);f0.add(0);}
		for(i=0;i<N;i++){e0.add(1);f0.add(1);}
		int[] e=new int[N*2+1];
		int[] f=new int[N*2+1];
		do{
			for(i=0;i<N*2;i++)e[i+1]=e[i]+e0.get(i);
			do{
				for(i=0;i<N*2;i++){
					f[i+1]=f[i]+f0.get(i);
					if(e[i]==f[i]&&e[i+1]==f[i+1])break;
				}
				if(i==N*2)r++;
			}while(next_permutation(f0));
		}while(next_permutation(e0));
		System.out.println(r);
	}
}

File: thisweek_masuipeo/q608/tyama_codeiq608_next.rs
fn reverse<T>(a:&mut[T],start_:usize,size:usize){
	let mut start=start_;
	let mut end=start+size-1;
	while start<end {
		{
			let p1: *mut T = &mut a[start];
			let p2: *mut T = &mut a[end];
			unsafe {p1.swap(p2);}
		}
		end-=1;
		start+=1;
	}
}

fn next_permutation<T: std::cmp::PartialOrd>(a:&mut[T],n:usize)->bool{
	if n<0||a.len()<n {return false}
	reverse(a,n,a.len()-n);
	let mut i=a.len()-2;
	loop {
		if a[i]<a[i+1] {break}
		if i==0 {
			reverse(a,0,a.len());
			return false;
		}
		i-=1;
	}
	let k=i;
	i=a.len()-1;
	while i>=k+1 {
		if a[k]<a[i] {break}
		i-=1;
	}
	let l=i;
	{
		let p1: *mut T = &mut a[k];
		let p2: *mut T = &mut a[l];
		unsafe {p1.swap(p2);}
	}
	reverse(a,k+1,a.len()-(k+1));
	return true;
}

fn main(){
	let N=6;
	let mut e0=vec![0 as i64; N*2];
	let mut f0=vec![0 as i64; N*2];
	let mut i=0;
	let mut r=0;
	while i<N {
		e0[N+i]=1;
		f0[N+i]=1;
		i+=1;
	}
	e0.sort();
	f0.sort();
	let mut e=vec![0 as i64; N*2+1];
	let mut f=vec![0 as i64; N*2+1];
	loop {
		i=0;
		while i<N*2 {
			e[i+1]=e[i]+e0[i];
			i+=1;
		}
		loop {
			i=0;
			while i<N*2 {
				f[i+1]=f[i]+f0[i];
				if e[i]==f[i]&&e[i+1]==f[i+1] {break}
				i+=1;
			}
			if i==N*2 {r+=1;}
			if !next_permutation::<i64>(&mut f0,N*2) {break}
		}
		if !next_permutation::<i64>(&mut e0,N*2) {break}
	}
	println!("{}",r);
}
